searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp

```diff
@@ -210,5 +228,96 @@ TEST("requireThatFastSearchLocationTermsWork") {
     EXPECT_TRUE(search(node, attribute_manager));
 #endif
 }
+
+AttributeVector::SP
+make_tensor_attribute(const vespalib::string& name, const vespalib::string& tensor_spec)
+{
+    Config cfg(BasicType::TENSOR, CollectionType::SINGLE);
+    cfg.setTensorType(ValueType::from_spec(tensor_spec));
+    return AttributeFactory::createAttribute(name, cfg);
+}
+
+AttributeVector::SP
+make_int_attribute(const vespalib::string& name)
+{
+    Config cfg(BasicType::INT32, CollectionType::SINGLE);
+    return AttributeFactory::createAttribute(name, cfg);
+}
+
+template <typename BlueprintType>
+const BlueprintType*
+as_type(const Blueprint& blueprint)
+{
+    const auto* result = dynamic_cast<const BlueprintType*>(&blueprint);
+    ASSERT_TRUE(result != nullptr);
+    return result;
+}
+
+class NearestNeighborFixture {
+public:
+    MyAttributeManager mgr;
+    vespalib::string attr_name;
+    AttributeContext attr_ctx;
+    FakeRequestContext request_ctx;
+    AttributeBlueprintFactory source;
+
+    NearestNeighborFixture(AttributeVector::SP attr)
+        : mgr(attr),
+          attr_name(attr->getName()),
+          attr_ctx(mgr),
+          request_ctx(&attr_ctx),
+          source()
+    {
+    }
+    void set_query_tensor(const TensorSpec& tensor_spec) {
+        request_ctx.set_query_tensor("query_tensor", tensor_spec);
+    }
+    Blueprint::UP create_blueprint() {
+        query::NearestNeighborTerm term("query_tensor", attr_name, 0, Weight(0), 7);
+        return source.createBlueprint(request_ctx, FieldSpec(attr_name, 0, 0), term);
+    }
+};
+
+TEST_F("nearest neighbor blueprint is created by attribute blueprint factory",
+       NearestNeighborFixture(make_tensor_attribute(field, "tensor(x[2])")))
+{
+    TensorSpec dense_x_2 = TensorSpec("tensor(x[2])").add({{"x", 0}}, 3).add({{"x", 1}}, 5);
+    f.set_query_tensor(dense_x_2);
+
+    auto result = f.create_blueprint();
+    const auto* nearest = as_type<NearestNeighborBlueprint>(*result);
+    EXPECT_EQUAL("tensor(x[2])", nearest->get_attribute_tensor().getTensorType().to_spec());
+    EXPECT_EQUAL(dense_x_2, DefaultTensorEngine::ref().to_spec(nearest->get_query_tensor()));
+    EXPECT_EQUAL(7u, nearest->get_target_num_hits());
+}
+
+void
+expect_empty_blueprint(AttributeVector::SP attr, const TensorSpec& query_tensor, bool insert_query_tensor = true)
+{
+    NearestNeighborFixture f(attr);
+    if (insert_query_tensor) {
+        f.set_query_tensor(query_tensor);
+    }
+    auto result = f.create_blueprint();
+    as_type<EmptyBlueprint>(*result);
```

Comment:
this testing is very indirect and confusing. If this is a use-case, consider doing just casting in the as_type function and asserting/expecting the result outside

Reply:
Fixed